File: Python/String/10_all_permutations.py
"""
Idea:
using backtracking , find all possible permutations of string
"""


class Solution:
    def all_permutation(self, input_str, low, high, st):

        # base case
        if low > high:
            st.add(input_str)
            return

        for i in range(low, high + 1):
            # at each iteration, swapping two characters , and calling the function again

            input_str_list = list(input_str)
            input_str_list[low], input_str_list[i] = input_str_list[i], input_str_list[low]
            new_input = ''.join(input_str_list)

            self.all_permutation(new_input, low + 1, high, st)

            input_str_list[low], input_str_list[i] = input_str_list[i], input_str_list[low]  # backtrack

    def find_permutation(self, S):
        st = set()  # using set for any duplicates
        ans = []
        self.all_permutation(S, 0, len(S) - 1, st)

        for item in st:
            ans.append(item)

        return ans


if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        S = input()
        ob = Solution()
        ans = ob.find_permutation(S)
        ans.sort()
        for i in ans:
            print(i, end=" ")
        print()

File: C++/String/10_all_permutations.cpp

/*
Idea:
using backtracking , find all possible permutations of string
*/
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution
{
public:

    void AllPermutation(string& input, int low, int high, set<string>& st)
    {
        // base case
        if (low > high)
        {
            st.insert(input);
            return;
        }
        for (int i = low;i <= high;++i)
        {
            // at each iteration, swapping two characters , and calling the function again
            swap(input[low], input[i]);
            AllPermutation(input, low + 1, high, ans);
            swap(input[low], input[i]);
        }

    }
    vector<string>find_permutation(string S)
    {
        set<string>st; //using set for any duplicates
        vector<string>ans;
        AllPermutation(S, 0, S.size() - 1, st);
        for (auto it : st)
            ans.push_back(it);
        return ans;
    }
};



int main() {
    int t = 1;
    //cin >> t;
    while (t--)
    {
        string S = "ABC";
        // cin >> S;
        Solution ob;
        vector<string> ans = ob.find_permutation(S);
        // sort(ans.begin(), ans.end());
        for (auto i : ans)
        {
            cout << i << " ";
        }
        cout << "\n";
    }
    return 0;
}

// } Driver Code Ends